modules/python/Makefile.env

```diff
@@ -1,76 +1,103 @@
-	ifeq ($(PY_CURRENT_VENV),)
-		PY_VENV:=.venv
+	# Unless WORKON_HOME is set, assume we're going to create ".venv" next to the Pipfile.
+	ifeq ($(WORKON_HOME),)
+		PIPENV := PIPENV_VENV_IN_PROJECT=1 $(PY_PIPENV)
 	else
-		PY_VENV:=$(PY_CURRENT_VENV)
+		PIPENV := PIPENV_VENV_IN_PROJECT= WORKON_HOME='$(WORKON_HOME)' $(PY_PIPENV)
 	endif
+	PIPENV_RUN := $(PIPENV) run
+	# Backwards compatible alias.
+	WITH_PIPENV := $(PIPENV_RUN)
 
-	PY_VENV_WORKDIR?=.
+	# Get the directory where the Pipfile is.
+	PIPENV_PIPFILE_DIR := $(shell $(PIPENV) --where 2> /dev/null)
 
-	ifeq ($(PY_PIPFILE_EXISTS),0)
-		ifeq ($(PY_VENV_WORKDIR),.)
-			PIPENV:=PIPENV_VENV_IN_PROJECT=1 $(PY_PIPENV)
-		else
-			PIPENV:=PIPENV_VENV_IN_PROJECT= WORKON_HOME='$(shell realpath $(PY_VENV_WORKDIR))' $(PY_PIPENV)
-			# if the venv doesn't exist then this will fail and PY_VENV will be empty
-			PY_VENV:=$(shell $(PIPENV) --venv 2> /dev/null)
-			ifneq ($(.SHELLSTATUS),0)
-				# means pipenv failed, so it doesn't exist yet. Need to cause a make env
-				_:=$(shell $(value PIPENV) install --dev --deploy)
+	ifneq ($(PIPENV_PIPFILE_DIR),)
+		# Check if the venv exists yet.
+		PY_VENV := $(shell $(PIPENV) --venv 2> /dev/null)
 
-				PY_VENV:=$(shell $(PIPENV) --venv 2> /dev/null)
+		ifneq ($(.SHELLSTATUS),0)
+			# The venv doesn't exist yet.
+			ifeq ($(WORKON_HOME),)
+				# Again, assume we're going to create ".venv" next to the Pipfile.
+				PY_VENV := $(PIPENV_PIPFILE_DIR)/.venv
+			else
+				# Create the venv, then check for it's location again.
+				_ := $(shell $(PIPENV_RUN) install --dev --deploy)
+				PY_VENV := $(shell $(PIPENV) --venv 2> /dev/null)
 			endif
 		endif
-
-		WITH_PIPENV:=$(PIPENV) run
 	endif
 endif
 
-Pipfile.lock:
-	@echo "You need to lock your environment!"
-
-$(PY_VENV)/bin/activate: Pipfile.lock
-	$(PIPENV) install --dev --deploy
-	if [ "$(PY_VENV_WORKDIR)" = "." ]; then touch $(PY_VENV)/bin/activate; fi
-
-$(PY_VENV): $(PY_VENV)/bin/activate
-
 .PHONY: pipenv/check
-pipenv/check: python/check ; $(call assert-set,PY_PIPENV) && echo "Found pipenv: $(PY_PIPENV)"
+## Test that pipenv is in PATH.
+pipenv/check: python/check
+	$(call assert-set,PY_PIPENV)
+	@echo "Found pipenv: $(PY_PIPENV)"
 
 .PHONY: pipenv/debug
+## Print build-harness vars about pipenv.
 pipenv/debug: python/debug
 	$(call print-var,PY_PIPENV)
-	$(call print-var,PY_PIPFILE_EXISTS)
-	$(call print-var,PY_CURRENT_VENV)
-	$(call print-var,PY_VENV)
+	$(call print-var,WORKON_HOME)
 	$(call print-var,PIPENV)
-	$(call print-var,WITH_PIPENV)
+	$(call print-var,PIPENV_PIPFILE_DIR)
+	$(call print-var,PY_VENV)
+	$(call print-var,PIPENV_RUN)
+
+$(PIPENV_PIPFILE_DIR)/Pipfile.lock: $(PIPENV_PIPFILE_DIR)/Pipfile
+	$(call assert-set,PIPENV_PIPFILE_DIR)
+	@$(MAKE) pipenv/lock
+
+$(PY_VENV)/bin/activate: $(PIPENV_PIPFILE_DIR)/Pipfile.lock
+	$(call assert-set,PIPENV_PIPFILE_DIR)
+	$(call assert-set,PY_PIPENV)
```

Comment:
Remove this assert-set since `PIPENV_PIPFILE_DIR` is checked on the line before and will only exist if `PY_PIPENV` existed?